src/core/api.c

```diff
@@ -616,6 +621,204 @@ mender_api_publish_deployment_status(const char *id, mender_deployment_status_t
         ret = MENDER_FAIL;
     }
 
+END:
+
+    /* Release memory */
+    mender_free(response);
+    mender_free(path);
+    mender_free(payload);
+    cJSON_Delete(json_payload);
+
+#ifdef CONFIG_MENDER_DEPLOYMENT_LOGS
+    /* Do this after we have released memory above, potentially giving us some
+       extra room we may need. */
+    if ((MENDER_OK == ret) && (MENDER_DEPLOYMENT_STATUS_FAILURE == deployment_status)) {
+        /* Successfully reported a deployment failure, upload deployment
+           logs.  */
+        if (MENDER_OK != mender_api_publish_deployment_logs(id)) {
+            mender_log_error("Failed to publish deployment logs");
+        }
+    }
+#endif /* CONFIG_MENDER_DEPLOYMENT_LOGS */
+
+    return ret;
+}
+
+#ifdef CONFIG_MENDER_DEPLOYMENT_LOGS
+static void
+append_depl_log_msg(char *msg, void *ctx) {
+    assert(NULL != ctx);
+
+    char  *tstamp   = NULL;
+    char  *level    = NULL;
+    char  *log_msg  = NULL;
+    cJSON *json_msg = NULL;
+    cJSON *messages = ctx;
+
+    /* Example log message we expect:
+     *   "[00:39:06.746,000] <err> mender: Unable to perform HTTP request"
+     * The code below goes through the string, searches for the expected parts and breaks the string
+     * down accordingly.
+     */
+
+    /* Start by setting log_msg to the whole message. In case all of the
+       break-down below fails, we send the whole message as the log message with
+       no extra metadata. */
+    log_msg = msg;
+
+    char *c = msg;
+    if ('[' == *c) {
+        /* if it does start with a timestamp, like above, store the pointer and find its end */
+        c++;
+        tstamp = c;
+        while (('\0' != *c) && (']' != *c)) {
+            c++;
+        }
+        if ('\0' == *c) {
+            goto DONE_PARSING;
+        }
+        *c = '\0';
+        c++;
+    }
+
+    if (' ' == *c) {
+        /* skip the space */
+        c++;
+    }
+
+    if ('<' == *c) {
+        /* if the log level follow, like above, store the pointer and find its end */
+        c++;
+        level = c;
+        while (('\0' != *c) && ('>' != *c)) {
+            c++;
+        }
+        if ('\0' == *c) {
+            goto DONE_PARSING;
+        }
+        *c = '\0';
+        c++;
+    }
+
+    if (' ' == *c) {
+        /* skip the space */
+        c++;
+    }
+
+    if ('\0' != *c) {
+        log_msg = c;
+        if (mender_utils_strbeginswith(log_msg, "mender: ")) {
+            log_msg += strlen("mender: ");
+        }
+    }
```

Comment:
I haven't reviewed the parsing, sorry. A unit tests will build up my confidence :)